fendermint/vm/interpreter/src/fvm/state/priority.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use crate::fvm::FvmMessage;
+use fvm_shared::bigint::BigInt;
+use fvm_shared::econ::TokenAmount;
+use num_traits::{ToPrimitive, Zero};
+
+/// The transaction priority calculator. The priority calculated is used to determine the ordering
+/// in the mempool.
+#[derive(Clone, Debug)]
+pub struct TxnPriorityCalculator {
+    /// Ring buffer of base fee history
+    base_fee_history: Vec<Option<TokenAmount>>,
+    /// Next slot in the ring buffer
+    next_slot: usize,
+}
+
+impl TxnPriorityCalculator {
+    pub fn new(size: usize) -> Self {
+        let mut v = Vec::with_capacity(size);
+        for _ in 0..size {
+            v.push(None);
+        }
+        Self {
+            base_fee_history: v,
+            next_slot: 0,
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.base_fee_history.len()
+    }
+
+    pub fn base_fee_updated(&mut self, base_fee: TokenAmount) {
+        self.base_fee_history[self.next_slot] = Some(base_fee);
+        self.next_slot = (self.next_slot + 1) % self.base_fee_history.len();
+    }
+
+    pub fn priority(&self, msg: &FvmMessage) -> i64 {
+        let base_fee = self.lowest_base_fee();
+
+        if msg.gas_fee_cap >= base_fee {
+            let i = msg.gas_fee_cap.clone() - base_fee + msg.gas_premium.clone();
+            i.atto()
+                .min(&BigInt::from(i64::MAX))
+                .to_i64()
+                .expect("clipped to i64 max")
+        } else {
+            0
+        }
```

Comment:
1. These kinds of conditionals are easier to follow when written as an early return. If the gas_fee_cap is below the base fee, return 0.
2. Are we sure 0 is the right value here? Why not return i64::MIN if the message is non-includable? Right now messages with gas_fee_cap == base_fee && gas_premium == 0 would rank exactly the same as non-includable messages. Seems wrong?
3. Adding the gas premium to the delta between the gas_fee_cap and the base fee is incorrect. The gas_fee_cap _includes_ the payable base fee **and** the gas premium. When the message is includable, our priority should be proportional to the **effective gas premium**. See how the [effective gas premium](https://github.com/filecoin-project/ref-fvm/blob/a8f9be18e76fcbe5d4b326d8c213eedb58121996/fvm/src/executor/default.rs#L91) is calculated in ref-fvm; copy the logic here.
4. i64::MAX atto is equal to 9.223372036854775807 base units. We are probably safe in clipping the effective premium to i64::MAX, since nobody sane will set such a high premium, but I'm uncomfortable with the panic here. I suggest this instead, which saturates to i64::MAX:

	```rust
	effective_premium.atto().to_i64().unwrap_or(i64::MAX);
	```